### ISSUE TITLE
Sharing GC.dll for NativeAOT executables

### ISSUE THREAD

**Issue by kant2002** [dotnet/runtime#75964]
From gitter
@hez2010 
> Is it possible to share a single GC among multiple nativeaot executables?
For example, can two binaries a.exe and b.exe which are compiled by nativeaot use a shared gc, so that the gc code won't be duplicated in each exe binary?

@MichalStrehovsky 
> It's not currently possible with shipped bits. But It is how things were set up in for .NET Native: there was an extra dll (mrt100_app.dll) that had the GC. in theory you should be able to still build that configuration from source - build https://github.com/dotnet/runtime/tree/main/src/coreclr/nativeaot/Runtime.Base/src into a shared library - this shared library will include the GC+the base services provided by the runtime. set InPlaceRuntime in CoreLib to false (https://github.com/dotnet/runtime/tree/main/src/coreclr/nativeaot/System.Private.CoreLib/src). then build an app as usual but don't pass Runtime.lib to linker and instead pass import library for the DLL you created in step 1 (we no longer need Runtime.lib because the DLL will provide those services, and we no longer need InPlaceRuntime parts of CoreLib because the DLL with provide those services too)

And now me "This sounds like fun project for hacktoberfest . Is this something which would be valuable for mainline?"

So would like to hear feedback, is this is desirable change from MS perspective?

**Comment by jkotas** [comment]
We took number of shortcuts over the years that are breaking this configuration. You would need to fix these shortcuts. For example, You would need to teach some places in the compiler to go through extra indirections in this config.

The GC code is relatively small to the rest of CoreLib, so the sharing benefits you get from this is not that great. .NET Native for UWP also compiled large part of the BCL into SharedLibrary.dll (look for SharedLibrary.dll in https://nuget.info/packages/Microsoft.Net.Native.SharedLibrary-x64/). It is where vast majority of the sharing benefits were, but the setup to make that work is very non-trivial.

**Comment by MichalStrehovsky** [comment]
One place that could potentially benefit from restoring building Runtime.Base as a separate library (instead of being folded into CoreLib) is type loader, if we were to move it to depend on Runtime.Base instead of CoreLib. I'm not sure how feasible that would be, but it would avoid an entire class of bugs like #70538 - I don't know how many stress bugs like that we have. Having a working Runtime.Base configuration would be a prerequisite to that. I see the biggest potential value in that.

**Comment by CurtHagenlocher** [comment]
My assumption when I hear "use a shared GC" is not (just) that the executable code itself is shared, but that there's a shared heap across several AOT-compiled components instead of multiple independent heaps. At least naively (which is about how well I understand this stuff) it seems like that would improve overall memory utilization for the process. Is that misguided? Are there fundamental problems with two different AOT runtimes sharing a heap?

**Comment by MichalStrehovsky** [comment]
Yes, this could mean a shared GC heap, not just the GC code.

> Are there fundamental problems with two different AOT runtimes sharing a heap?

The native AOT runtime was originally designed for this, but we haven't built this way for a long time. We'd probably need tweaks to the original layering too (e.g. only share things like GC/EH/suspension, not things like casting and interface dispatch that were originally also shared). It requires quite a bit of architectural thought.

**Comment by adams85** [comment]
Beyond reduced memory utilization, shared GC would be pretty much necessary for applications that want to implement a plugin system. At least, having one GC instance per plugin doesn't sound like a good idea. (Of course, you could implement the plugins in a non-GC language, but who would want to do that once we have C#... 😄)

So I think this is a shortcoming that still prevents NativeAOT from being an all-around solution for application development. (BTW, kudos to you, guys, NativeAOT already kicks ass in its current form! Love that I can compile my apps to native code and they start _instantaneously_. This is an experience which we've gotten pretty unused to lately.)

**Comment by MichalPetryka** [comment]
> Beyond reduced memory utilization, shared GC would be pretty much necessary for applications that want to implement a plugin system. At least, having one GC instance per plugin doesn't sound like a good idea. (Of course, you could implement the plugins in a non-GC language, but who would want to do that once we have C#... 😄)

If you want C# plugins in your app then you likely shouldn't be using AOT compilation.

This makes more sense for non C# applications that use arbitrary native plugins.

**Comment by adams85** [comment]
> If you want C# plugins in your app then you likely shouldn't be using AOT compilation.

Problem is that JIT-compiled .NET apps have a significant startup time. When developing, e.g., a desktop app, I prefer to AOT-compile it for fast startup. And in case I want that app pluggable, then I have the situation I mentioned in my previous comment. Of course, I'm not saying it's a super common use case. It just would be awesome if it were possible.

> This makes more sense for non C# applications that use arbitrary native plugins.

Now, this is a more common use case for sure.

**Comment by MichalPetryka** [comment]
> Problem is that JIT-compiled .NET apps have a significant startup time. When developing, e.g., a desktop app, I prefer to AOT-compile it for fast startup.

Have you looked into [ReadyToRun](https://learn.microsoft.com/en-us/dotnet/core/deploying/ready-to-run)? That's the intended partial AOT scenario for fast startup.

**Comment by adams85** [comment]
No, I'm not really familiar with ReadyToRun. Thanks for bringing it to my attention. Looks like it covers the use case I outlined above, but I can imagine that there are still cases where fully native code is preferred. 

However, let's not dive into this deeper. I think we already identified some real use cases for the shared GC model.
